src/app/modules/submission/encrypt-submission/encrypt-submission.controller.ts

```diff
@@ -647,13 +663,219 @@ const submitEncryptModeForm: ControllerHandler<
+ */
+const getEncryptedAttachmentsMapFromAttachmentsMap = async (
+  attachmentsMap: Record<string, Buffer>,
+  publicKey: string,
+): Promise<StorageModeAttachmentsMap> => {
+  const attachmentPromises = Object.entries(attachmentsMap).map(
+    ([id, attachment]) => encryptAttachment(attachment, { id, publicKey }),
+  )
+
+  return Promise.all(attachmentPromises).then((encryptedAttachmentsMeta) =>
+    chain(encryptedAttachmentsMeta)
+      .keyBy('id')
+      // Remove id from object.
+      .mapValues((v) => omit(v, 'id'))
+      .value(),
+  )
+}
+
+/**
+ * Reference implementation taken from frontend/src/features/public-form/utils/createSubmission.ts for purpose of demonstrating shifted encryption boundary.
+ * Utility to filter out responses that should be sent to the server. This includes:
+ * 1. Email fields that have an autoreply enabled.
+ * 2. Verifiable fields to verify its signature on the backend.
+ */
+const filterSendableStorageModeResponses = (
+  formFields: FormFieldDto[],
+  responses: FieldResponse[],
+) => {
+  const mapFieldIdToField = keyBy(formFields, '_id')
+  return responses
+    .filter((r): r is EmailResponse | MobileResponse => {
+      switch (r.fieldType) {
+        case BasicField.Email: {
+          const field = mapFieldIdToField[r._id]
+          if (!field || field.fieldType !== r.fieldType) return false
+          // Only filter out fields with auto reply set to true, or if field is verifiable.
+          return field.autoReplyOptions.hasAutoReply || field.isVerifiable
+        }
+        case BasicField.Mobile: {
+          const field = mapFieldIdToField[r._id]
+          if (!field || field.fieldType !== r.fieldType) return false
+          return field.isVerifiable
+        }
+        default:
+          return false
+      }
+    })
+    .map((r) => pick(r, ['fieldType', '_id', 'answer', 'signature']))
+}
+
+/**
+ * Encrypt submission content before saving to DB.
+ */
+const encryptSubmission: ControllerHandler<
+  { formId: string },
+  SubmissionResponseDto | SubmissionErrorDto,
+  ParsedStorageModeSubmissionBody
+> = async (req, res, next) => {
+  const { formId } = req.params
+
+  const logMeta = {
+    action: 'convertToEncryptMode',
+    ...createReqMeta(req),
+    formId,
+  }
+
+  // Retrieve form definition. TODO: extract this step when validation is added.
+  const formResult = await FormService.retrieveFullFormById(formId)
+  if (formResult.isErr()) {
+    logger.warn({
+      message: 'Failed to retrieve form from database',
+      meta: logMeta,
+      error: formResult.error,
+    })
+    const { errorMessage, statusCode } = mapRouteError(formResult.error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+
+  // Guardrail to prevent new endpoint from being used for regular storage mode forms.
+  // TODO (FRM-1232): remove this guardrail when encryption boundary is shifted.
+  if (!formResult.value.get('newEncryptionBoundary')) {
```

Comment:
I'm weighing this against using a feature flag. With this new column, I will need to update all existing records' value to `true` for the boundary shift. And after the boundary shift, I would be storing a value that will no longer be used in all form definitions.

Reply:
No ah we can just remove this guardrail, then this endpoint will be available to all storage mode forms. But yes, this implementation will not provide any flexibility in terms of controlling the percentage rollout and there is no way of updating this without going into the DB. I was thinking maybe we can do a feature flag system in a separate PR maybe? Since this PR is already pretty sizeable (although it's mostly refactoring).

Reply:
Adding on to the discusison - i thought the plan was to do it only for forms with attachment first. in that case maybe the 'guardrail' can come in the form of a list of field types which are allowed.

which is a separate discussion from phased rollout. for phased rollout, i thought it is more of a frontend flag rather than backend flag (similar to what we did for react). because it is a phased rollout in terms of % of respondents, not in terms of % of forms.

Reply:
I think there is a need to protect the backend endpoint itself cos this endpoint opens up an easier way to send malicious attachments. So this is slightly different from our other features which are usually just blocked on our frontend since there's no harm using those endpoints independently.

Actually to clarify, so the attachment-first approach is to showcase our new virus scanning capabilities or? Why would we open for forms with attachment fields first before the others?

Reply:
> Actually to clarify, so the attachment-first approach is to showcase our new virus scanning capabilities or? Why would we open for forms with attachment fields first before the others?

I think this was to achieve phased rollout (see the [RFC](https://docs.google.com/document/d/1VmNXS_xYY2Yg30AwVqzdndBp5yRJGSDsyjBnH51ktyc/edit) - Phase 1). Because virus scanning was the main benefit so we would start with forms with attachment fields. If there are any bugs / problems, it would be scoped to forms with attachment fields (might be easier to contact admins to remediate)

Reply:
Ahh yes okay, sorry I missed that. I presume we could add the flag for forms with attachment fields first for the first 3 months to track numbers and limit the scope of any bugs. But this should only be out after virus scanning is in place!

Reply:
but as [discussed on slack](https://opengovproducts.slack.com/archives/CKHLS3W3X/p1691113708282249?thread_ts=1691037615.319729&cid=CKHLS3W3X), I think for this particular PR, we should still have this guardrail instead of enabling it for certain forms immediately!